openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py

```diff
@@ -0,0 +1,236 @@
+        should_continue: boolean indicating whether ingestion should continue after
+                         this batch has been processed
+        """
+        batch = None
+        should_continue = True
+
+        try:
+            # Get the API response
+            response_json = self.get_response_json(query_params)
+
+            # Build a list of records from the response
+            batch = self.get_batch_data(response_json)
+
+            # Optionally, apply some logic to the response to determine whether
+            # ingestion should continue or if should be short-circuited. By default
+            # this will return True and ingestion continues.
+            should_continue = self.get_should_continue(response_json)
+
+        except Exception as e:
+            logger.error(f"Error due to {e}")
+
+        return batch, should_continue
+
+    def get_response_json(self, query_params: Dict):
+        """
+        Make the actual API requests needed to ingest a batch. This can be overridden
+        in order to support APIs that require multiple requests, for example.
+        """
+        return self.delayed_requester.get_response_json(
+            self.endpoint, self.retries, query_params, headers=self.headers
+        )
+
+    def get_should_continue(self, response_json):
+        """
+        This method should be overridden when an API has additional logic for
+        determining whether ingestion should continue. For example, this
+        can be used to check for the existence of a `continue_token` in the
+        response.
+        """
+        return True
+
+    @abstractmethod
+    def get_batch_data(self, response_json):
+        """
+        Take an API response and return the list of records.
+        """
+        pass
+
+    def process_batch(self, media_batch):
+        """
+        Process a batch of records by adding them to the appropriate MediaStore.
+        Returns the total count of records ingested up to this point, for all
+        media types.
+        """
+        record_count = 0
+
+        for record in media_batch:
+            record_data = self.get_record_data(record)
+
+            if record_data is not None:
+                record_count += 1
+
+                # We need to know what type of record we're handling in
+                # order to add it to the correct store
+                media_type = self.get_media_type(record)
+
+                # Get the store for that media_type
+                store = self.media_stores[media_type]
+                store.add_item(**record_data)
+
+        return record_count
+
+    @abstractmethod
+    def get_media_type(self, record):
+        """
+        For a given record, return the media type it represents (eg "image", "audio",
+        etc.) If a provider only supports a single media type, this may be hard-coded
+        to return that type.
+        """
+        pass
+
+    @abstractmethod
+    def get_record_data(self, record):
```

Comment:
One more consideration:
Science museum has 'Objects' that can have several images described inside the `multimedia` dictionary. What is the best way of handling such nested items?

Reply:
Maybe 
`record_data = self.get_record_data(record)`
could return a list of items?

Reply:
Oh, that's a good point -- SMK also does something similar, with `alternative_images`. 👀

Reply:
Hm, I think the vast majority of providers will be returning information for a single record. I pushed a change in [6112969](https://github.com/WordPress/openverse-catalog/pull/555/commits/61129690bee1a72a8587c226516af925d0c8274f) that allows `get_record_data` to return a single dict _or_ a list, but I'm not 100% sold. How do you feel about that change, or do you think it would be better to just require all providers to return a list here?

Reply:
This is exactly what I wanted to suggest! For the same reasons as you state: it doesn't make sense to make it a list only because a couple out of all providers require it, it is better to handle it within the `ingester` class.
In fact, I added similar changes locally, and my refactored Science museum script works, and it is so much cleaner!!! It works well except for one small piece that I'll describe in a different comment.